### PR TITLE
Barbering Menu TGUI (with previews!)

### PR DESCRIPTION
**Pull request by CaioFrancisco**
## About the PR 

This PR updates the old Barbering Menu that simply listed hair names into a shiny, new TGUIzed version. Here are some screenshots showcasing the new UI:

Screenshot_527
Screenshot_528
Screenshot_529

## Why's this needed? 

This UI is much, _much_, **much**, _**MUCH**_ better than the previous one, and making Barber more viable is always better. Though, I have some worries about how I coded it: it iterates through a large list of haircuts and has to go through some expensive processes of turning it's icon into base64 for it to properly open. Other than that, the code is as non-expensive as I tried to make it. If there are any glaring optimization opportunities, please ping me on Discord for me to respond as quickly as I can: Caio029.

## Changelog 

```changelog
(u)Caio029
(+)Updates the Barbering Menu into a new, fancy TGUI version with previews!
```

### CODE REVIEW

**Comment by Retrino** [comment]
this seems pretty clunky without a search function. the previews are good but with the amount of hairs we have it seems anti qol to remove that feature

**Comment by CaioFrancisco** [comment]
Oh yeah, the previous TGUI of the list of hairs had that feature. I'll add that.

**Comment by CaioFrancisco** [comment]
Oh christ, I've found a critical glitch in the PR. Because of how items are mirrored from the Barbee to the character preview doll, by deleting the character preview doll, you delete the barbee's items. I'll fix that and add a search bar in my PR, and as soon as that's done, I'll set it back ready for review.

**Comment by CaioFrancisco** [comment]
The PR should be good enough for a code review now.

**Comment by CaioFrancisco** [comment]
!merge_upstream

**Comment by CaioFrancisco** [comment]
???????????????
Screenshot_413

The branch has no merge conflicts, so I'll go ahead and remove the merge conflict label.

**Comment by CaioFrancisco** [comment]
There were no git-related crimes before this point.

**Comment by CaioFrancisco** [comment]
Look, I'm getting tired of having to fix merge conflicts and/or fixing my branch after screwing up everything. I'll merge and re-compile the bundle once someone does a code review.

**Comment by ZeWaka** [comment]
Nor do you need to fix the merge conflicts with tgui files

File: code/datums/character_preview.dm
Comment (mordent-goonstation) [inline code/datums/character_preview.dm:109]: This comment is somewhat concerning. Could you expand on it (i.e. the "why")?

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:37]: Here's the window being stretched as long as TGUI allows me to:
Screenshot_430

And here's the window being narrowed to the maximum without breaking the text structure due to being too narrow:
Screenshot_431

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js]: I removed it from the preview buttons, but otherwise, I felt like it added a nice touch to look at.

File: code/datums/character_preview.dm
Comment (CaioFrancisco) [reply code/datums/character_preview.dm]: I tried to explain why it's not safe to use this proc. Tell me if I should change the warning.

**Comment by CaioFrancisco** [comment]
!merge_upstream

File: code/datums/character_preview.dm
Comment (mordent-goonstation) [inline code/datums/character_preview.dm:109]: This comment is not useful to the next coder, imo. Is there actually a problem with using the proc? If so, don't make it and do it the way you would recommend. If there isn't a problem, is the comment needed?

File: code/datums/components/barber.dm
Comment (mordent-goonstation) [inline code/datums/components/barber.dm:477]: Do `start_and_wait` actions do delay stuff? If so, probably need to check continued existence/validity of the various things we're acting on below?

File: code/datums/components/barber.dm
Comment (mordent-goonstation) [inline code/datums/components/barber.dm:497]: Rather than make this every time, and in `ui_static_data`, would it make sense to create when the datum is initialized and refer to here?

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:35]: Not a fan of two items with percentages totaling 97%. Implies layout hacks to me a little. I get that this item is `position: fixed` so it doesn't affect the layout of the earlier one. Seems like these aren't `Stack.Item`s at this point, perhaps?

I believe the `PreviewWindow` could be a fixed width; would using a fixed width prevent distortion issues as you increase/decrease width?

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:51]:
```suggestion
  return filtered_list.map((hairName) => (<HairPreview
    key={all_hair_styles[hairName].id}
    hair_style={all_hair_styles[hairName]}
    hair_name={hairName} />));
```

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:79]: A `Stack` should probably have `Stack.Item`s as children, rather than random other things (`Button`s, in this case). While I'm sure it works, I believe it's unintended behavior. Not 'defined' behavior, at least.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:89]:
```suggestion
```

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:90]: See other comments re: `Stack` and `Stack.Item`.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:107]: Apply `mb` to `Stack` rather than individual `Box` elements? Also sounds like you *might* want the `Section` itself to have a `Stack` as its immediate child, possibly? Should save some of the manual styling being done.

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:35]: I'll have to admit, this is indeed somewhat layout hacks. If I try to use up the other 3%, the `PreviewWindow` will get over the scroll bar. Although, I also don't feel comportable using fixed widths, since it will actually worsen distortion by re-sizing the window, because it forces the `PreviewWindow` to get over the scroll bar once again if there's too little space on the window. This is, as far as I know, the best solution to this issue.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:79]: I attempted to cover most of the elements under `<Stack.Item>` where it should be. Tell me if it's good or nah, or if I was excessive when turning everything into `<Stack.Item>`

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:106]: It wasn't working for whatever reason, but after nesting it under a `<Stack.Item>`, it looked neat enough to not need any margins. Here's how it looks right now:
Screenshot_460

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:67]: No need for this. If you do want it as a visual thing, use a border on one of the adjacent `Stack.Item`s instead.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:112]: Use of `width="100%"` on one of the items and looking at an earlier screenshot has me thinking this is a vertical `Stack`?

https://github.com/goonstation/goonstation/blob/master/tgui/docs/component-reference.md#stack

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:120]:
```suggestion
      <LabeledList>
        <LabeledList.Item label="Top Hair">{current_hair_style["top"]}</LabeledList.Item>
        <LabeledList.Item label="Middle Hair">{current_hair_style["middle"]}</LabeledList.Item>
        <LabeledList.Item label="Bottom Hair">{current_hair_style["bottom"]}</LabeledList.Item>
      </LabeledList>
```
Untested, but this seems like a labeled list to me? https://github.com/goonstation/goonstation/blob/master/tgui/docs/component-reference.md#labeledlist

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:137]:
```suggestion
      <Stack vertical>
        <Stack.Item>
          <Button icon="caret-up" onClick={() => act("update_preview", { "action": "change_direction", "direction": "north" })} />
        </Stack.Item>
        <Stack.Item>
          <Stack justify="space-around">
            <Stack.Item>
              <Button icon="caret-left" onClick={() => act("update_preview", { "action": "change_direction", "direction": "west" })} />
            </Stack.Item>
            <Stack.Item>
              <Button icon="caret-right" onClick={() => act("update_preview", { "action": "change_direction", "direction": "east" })} />
            </Stack.Item>
          <Stack>
        </Stack.Item>
        <Stack.Item>
          <Button icon="caret-down" onClick={() => act("update_preview", { "action": "change_direction", "direction": "south" })} />
        </Stack.Item>
      </Stack>
```
Something like this? Not tested, trying to simplify a lot of the weird `Stack` usage.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:87]:
```suggestion
      <Stack vertical>
        <Stack.Item>
          <Image pixelated width="60px" height="100px" src={`${hair_style["hair_icon"]}`} />
        </Stack.Item>
        <Stack.Item>
          <Box fontSize="15px" textAlign="center">{hair_name}</Box>
        </Stack.Item>
        <Stack.Item>
          <Button color="blue" height="20px" icon="cut" onClick={() => act("do_hair", { "style_id": hair_style["hair_id"] })}>Cut</Button>
          <Button color="blue" height="20px" icon="eye" onClick={() => act("update_preview", { "action": "new_hair", "style_id": hair_style["hair_id"] })}>Preview</Button>
        </Stack.Item>
      </Stack>
```
Possibly something like this (untested). Might want to make use of `Flex` if you want tighter spacing.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:71]: Can probably be written simply using `Flex` and `Flex.Item` around each thing, rather than having one item for the buttons that have hardcoded mx, one for the Create Wig button, and one for the (unnecessary?) fake border.

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
I took another pass at the various layout stuff. Links to relevant docs:

* https://github.com/goonstation/goonstation/blob/master/tgui/docs/component-reference.md#stack
* https://github.com/goonstation/goonstation/blob/master/tgui/docs/component-reference.md#flex
* https://github.com/goonstation/goonstation/blob/master/tgui/docs/component-reference.md#labeledlist

If you get into a weird spot with it, let me know and I can try to pull to my local and take a stab at it. Please ping me (`@Mordent`) on Discord in #imcoder or somewhere if/when you do.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:120]: it works nicely, i'll add it in the next commit

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:137]: The "up" and "down" arrows get thrown to the left. Since `justify` doesn't seem to work on `vertical` Stacks and I'm not very knowledgeable about anything else to do the trick, I'll just leave this suggestion open and leave the spaghetti on my branch intact.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:87]: Seems to work nicely, I'll add it in the next commit.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:22]: I tried replacing `Stack` with either `Section` or `div`, but I couldn't find a way for both to do the same effect as `HairPreviewList` on one side and `PreviewWindow` on another.

**Comment by CaioFrancisco** [comment]
By the way, I had to go back to using percentages for `PreviewWindow` because the direction arrows weren't re-dimensioning when the window did, making most buttons inaccessible if the user ever made the window narrower.

**Comment by mordent-goonstation** [comment]
Running this locally, I get an error when attempting to press "Preview" on a hair style.

```
Exception has occurred: Overlays on "Yasmin Nazari" (/mob/living/carbon/human) [0x30000fc] were modified by non-UpdateOverlays method.
```

**Comment by CaioFrancisco** [comment]
@mordent-goonstation Merged from master and rebuilt bundle, as I usually do before diagnosing problems. I haven't runtimed after previewing on myself and 5 other human NPCs. Maybe the merge fixed it?

**Comment by mordent-goonstation** [comment]
Did the same locally, does seem to have fixed the issue I was seeing. Working on some of the layout stuff now, will put a branch on my fork.

**Comment by mordent-goonstation** [comment]
Problem noticed while messing with: stepping away from it doesn't affect whether it's interactive/updating. You can't actually cut someone's hair because of the action bar, but it's weird that it implies you can.

**Comment by CaioFrancisco** [comment]
> Problem noticed while messing with: stepping away from it doesn't affect whether it's interactive/updating. You can't actually cut someone's hair because of the action bar, but it's weird that it implies you can.

Oh, right, forgot to add flags for the actionbar. It's my first proper one.

**Comment by mordent-goonstation** [comment]
@CaioFrancisco I've put my changes up on my fork's branch so you can see the relevant commit (https://github.com/mordent-goonstation/goonstation/tree/shoutout_to_all_barbers_out_there). There was a change a little bit ago that has changed previews that means merging into master won't work right now, i.e. my branch is broken because I `master` in. If you want to see what I changed you can take a look at the single commit I did, it should blend nicely on top of your branch, i.e. should be cherry-pickable if you rebuild tgui after (https://github.com/mordent-goonstation/goonstation/commit/7f4cc89be0a37080a23cb42a7368cedeca7666cf).

- https://github.com/goonstation/goonstation/pull/12976 This is the PR that changed how preview objects work. May be as simple as renaming in your branch. You'll need to address that.

Sorry this took a bit, 101 RL things came up!

**Comment by CaioFrancisco** [comment]
Looks great to me. Fixed the few other issues, if you get one tile apart from the barbee, the window will close now. Thank you a lot for helping me out on this, Mordent.

**Review by mordent-goonstation — APPROVED** [review]
tgui side looks fine. Could use an updated screenshot in the main post. Leaving for another dev to look through the non-tgui side, the parts of my brain that remember how all of that works are non-responsive right now.

Will obviously need a tgui bundle rebuild before merging.

File: tgui/packages/tgui/interfaces/BarberingMenu.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/BarberingMenu.js:1]: @CaioFrancisco Could you toss a license header on here? Take a look at other interfaces if you're not sure. I guess probably throw me on it as well if you want, though I will not require that as I mostly just shuffled your stuff around.

Comment (CaioFrancisco) [reply tgui/packages/tgui/interfaces/BarberingMenu.js:1]: Right, I'll do it as soon as I can.

**Review by mordent-goonstation — APPROVED** [review]
Broadly approving. Would appreciate a dev looking at the way that the preview icons are rendered, I'm *sure* there's some jank there (because with those there always is).

Not requiring in this PR, but as a future change would be good to make the arrows rotate (similar to the player panel one) rather than set direction. This would make the UI there a little more compact and intuitive.

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:65]: This should definitely not be re-initialized on every component, make it static or something.

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:76]: if you're checking `istype(src)`, that means you should be moving the behaviour onto the child type (most of the time anyway)

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:536]: Same here, this should be handled with inheritance rather than a copy pasted snippet (and all the other instances of this being duplicated)

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:518]: and here too

**Comment by CaioFrancisco** [comment]
Sorry for the enormous delay, the PR is once again ready to be reviewed.

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:86]: ?

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:94]: you should be able to use `initial` to avoid that, like `initial(styles.id)`

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:112]: why not just have this as a var on the TYPEINFO and not repeat the icon init code?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Also upon testing, it seems the hairstyle gets reset on rotating the preview

File: code/datums/components/barber.dm
Comment (CaioFrancisco) [reply code/datums/components/barber.dm:112]: `concrete_typesof` is a proc, and you can't use procs when defining a var outside of a proc, which `TYPEINFO()` isn't. on an up-coming commit, i try to change a bit of the above

**Comment by CaioFrancisco** [comment]
by the way, if i don't define `TYPEINFO()` at all, `TYPEINFO_NEW` doesn't run.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Code looks better, preview is still broken. Currently you appear naked until any changes are made, at which point your face disappears.
image
Might want to look at how the character preferences menu does it?

**Comment by CaioFrancisco** [comment]
No merge conflict, I'll be removing the tag

**Comment by CaioFrancisco** [comment]
forgot to re-request review, sorry.

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [inline code/datums/components/barber.dm:563]: this is one massive memory leak, either don't replace them every time or at least `qdel` the old ones

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/datums/components/barber.dm
Comment (TobleroneSwordfish) [reply code/datums/components/barber.dm:563]: image

Comment (CaioFrancisco) [reply code/datums/components/barber.dm:563]: oh, right, shit i forgot that. that's one dumb mistake. sorry, i'll fix that.

**Comment by ThatFiveGuys** [comment]
I think itd be scope creepey to do it in this pr, but a variation of this for character setup would be wonderful

**Comment by CaioFrancisco** [comment]
> I think itd be scope creepey to do it in this pr, but a variation of this for character setup would be wonderful

I'd do that, but I've been doing this PR since last year, and I'd be much happier to do this in another PR or not doing it at all to give barbers some more well-deserved love.

**Comment by CaioFrancisco** [comment]
Is fixed, reworked a bit of the code to be safer, probably won't explode the RAM like before. (i am still astonished that i willingly committed that piece of code.)

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Okay, memory leak fixed, one last bug!
Failing to cut someone's hair while they have no hair results in a blank wig being spawned: 
image
Since this is only possible by keeping the window open after cutting all their hair off, maybe doing that should just close the window.

**Comment by CaioFrancisco** [comment]
tried to make code less shité as well. tell me if it needs any more work.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
looks like something is being disposed out of order here:
image
happens when you fail a barbering action

**Comment by CaioFrancisco** [comment]
I cannot, for the life of me, figure out when, where and how does `preview` get nullified.

**Pull request merged**
